doc/sphinx/language/gallina-specification-language.rst

```diff
@@ -1581,33 +1606,88 @@ using the keyword :cmd:`Qed`.
-between ``#[`` (hash and opening square bracket) and ``]`` (closing square bracket)
-and separated by commas ``,``. Multiple space-separated blocks may be provided.
+.. insertprodn all_attrs legacy_attrs
+
+.. prodn::
+   all_attrs ::= {* #[ {*, @attr } ] } {? @legacy_attrs }
+   attr ::= @ident {? @attr_value }
+   attr_value ::= = @string
+   | ( {*, @attr } )
+   legacy_attrs ::= {? {| Local | Global } } {? {| Polymorphic | Monomorphic } } {? Program } {? {| Cumulative | NonCumulative } } {? Private }
+
+Attributes modify the behavior of a command or tactic.
+Syntactically, most commands and tactics can be decorated with attributes, but
+attributes not supported by the command or tactic will be flagged as errors.
+
+The order of top-level attributes doesn't affect their meaning.  ``#[foo,bar]``, ``#[bar,foo]``,
+``#[foo]#[bar]`` and ``#[bar]#[foo]`` are equivalent.
 
-Each attribute has a name (an identifier) and may have a value.
-A value is either a :token:`string` (in which case it is specified with an equal ``=`` sign),
-or a list of attributes, enclosed within brackets.
+The legacy attributes (:n:`@legacy_attrs`) provide an older, alternate syntax
+for certain attributes.
 
 Some attributes are specific to a command, and so are described with
 that command. Currently, the following attributes are recognized by a
 variety of commands:
 
-``universes(monomorphic)``, ``universes(polymorphic)``
-    Equivalent to the ``Monomorphic`` and
-    ``Polymorphic`` flags (see :ref:`polymorphicuniverses`).
+.. attr:: universes(monomorphic)
+   :name: universes(monomorphic)
+
+   See :ref:`polymorphicuniverses`.
+
+.. attr:: Monomorphic
+
+   Legacy attribute equivalent to :attr:`universes(monomorphic)`.
+
+.. attr:: universes(polymorphic)
+   :name: universes(polymorphic)
+
+   See :ref:`polymorphicuniverses`.
+
+.. attr:: Polymorphic
+
+   Legacy attribute equivalent to :attr:`universes(polymorphic)`.
 
-``program``
-    Takes no value, equivalent to the ``Program`` flag
-    (see :ref:`programs`).
+.. attr:: program
 
-``global``, ``local``
-    Take no value, equivalent to the ``Global`` and ``Local`` flags
-    (see :ref:`controlling-locality-of-commands`).
+   See :ref:`programs`.
 
-``deprecated``
-    Takes as value the optional attributes ``since`` and ``note``;
-    both have a string value.
+.. attr:: Program
+
+   Legacy attribute equivalent to :attr:`program`.
+
+.. attr:: global
+
+   See :ref:`controlling-locality-of-commands`.
+
+.. attr:: Global
+
+   Legacy attribute equivalent to :attr:`global`.
+
+.. attr:: local
+
+   See :ref:`controlling-locality-of-commands`.
+
+.. attr:: Local
+
+   Legacy attribute equivalent to :attr:`local`.
+
+.. attr:: Cumulative
+
+   Legacy attribute THAT DOES WHAT?  Allowed only in a polymorphic context.
+
+.. attr:: NonCumulative
+
+   Legacy attribute THAT DOES WHAT?  Allowed only in a polymorphic context.
+
+.. attr:: Private
+
+   Legacy attribute that DOES WHAT?
```

Comment:
Private inductive types are an undocumented (yet critical to some users) feature. See #11223.